### DISCUSSION TITLE
MgrsLines for bounds and zoom level

### DISCUSSION THREAD

**Discussion by tercjak** [ngageoint/mgrs-java#3] [Ideas]
I need method to get Mgrs Grid Lines frm  bounds and zoom level
With this i can implement mgrs Grid lines in Mapbox .

**Comment by tercjak** [comment]
https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames
[Mgrs.zip](https://github.com/ngageoint/mgrs-java/files/12195854/Mgrs.zip)

**Comment by bosborn** [comment]
There is an [example in the README](https://github.com/ngageoint/mgrs-java#draw-tile-template).
Get the grids by the zoom level and the grid range by the bounds.

For a concrete example, [see the Android implementation](https://github.com/ngageoint/mgrs-android/blob/309d6a3c770d851ab4e565a4a4ba27c18d968c55/mgrs/src/main/java/mil/nga/mgrs/grid/style/Grids.java#L131).

**Reply by tercjak** [reply]
I have already dealt with this but the example is not adequate.

1 example draw to bitmap instead of adding lines to map
2 example is only for bitmaps OSM maps and Google Maps and not for vector Mapbox Maps .
3 Api has the following bug: After downloading lines from zoomGrids every 4th column has lines from the slant not even full . It helps to filter lines to meet the condition
it.point1.longitude <= it.point2.longitude
4 for GridType.GZD draws 4 lines instead of 2 actual lines needed
5 We already have bounds for zone Point southwest = getSouthwest();
		Point northwest = getNorthwest();
		Point northeast = getNortheast();
		Point southeast = getSoutheast();
but instead of using them we are in GridZone.getLines (198c line of code) we re-create them using UTM class to update them later.
6 The tree structure of zoomGrids is strange and causes duplication of lines.They are close to each other by eastingPrecision and northingPrecision 
GridZone.getLines(zoom,bounds)function should return all the lines in a given zone that are in the grid and not the tree but also it would be nice if you could pass a function to control with which precision level we drawing at which zoom.

**Reply by bosborn** [reply]
The main purpose of this library is for drawing tiles and performing MGRS coordinate operations.  When I pointed you to a Bitmap example, the idea was it can be duplicated with tile creation in the form of a BufferedImage.  The lines returned can be from multiple grid precisions for the purpose of drawing the different styles overlapping.  You could adjust the zoom levels for the grid precision to not overlap if that is your desire in Grids.java.  The lines returned are also in the precision level sections/chunks which is required for display/draw accuracy.  Any line overlapping bounds is also included for the purpose of tile drawing.  All 4 surrounding lines of a GZD are returned for drawing.  The tree of Grids is sorted by grid type precisions for drawing order (most precise / smallest precision meters first).  I'm sure you can adapt to use for actual lines on a map with some adjustments.  If you found an actual bug you can submit a test showing the issue with both wrong and expected results or submit a pull request.  The algorithm works well for our Android and iOS apps without issues and many of your issues are by design for tile drawing.

**Comment by tercjak** [comment]
>The algorithm works well for our Android and iOS apps without issues and many of your issues are by design for tile drawing.
No it has a bug GridZone.getLines returns diagonal lines ( it.point1.longitude <= it.point2.longitude) .
![Screenshot_20230801-125022](https://github.com/ngageoint/mgrs-java/assets/7704977/dde75a9c-1f0c-4bbe-8480-5fcb2c3df99c)

>All 4 surrounding lines of a GZD are returned for drawing.
That's right, but it looks the same as if you drew the vertical and horizontal lines alone for the zones ( nw-ne line of one zone will be sw-se line )of another and similarly ( nw-sw line of one zone will be sw-se line will be ne-se line)
top bound of one zone is bottom of other etc.

The algorithm that draws bitmap is also wrong because it does not display them. I am still puzzled by the lines leading to the triangle near Bydgoszcz.

**Reply by bosborn** [reply]
I'm not sure what you mean, but this is probably not the best resource if you are trying to learn how MGRS works.

![Screenshot_20230802-052954](https://github.com/ngageoint/mgrs-java/assets/11667761/09e23ec8-be21-430f-b938-d6365b68a975)
![Screenshot 2023-08-02 at 5 08 19 AM](https://github.com/ngageoint/mgrs-java/assets/11667761/38fff99f-c4b4-4d23-9a39-b3c3ea68ae3c)
![Screenshot 2023-08-02 at 5 00 51 AM](https://github.com/ngageoint/mgrs-java/assets/11667761/82a0524e-2495-4f9b-a22e-1c74648df248)
![Screenshot 2023-08-02 at 4 52 06 AM](https://github.com/ngageoint/mgrs-java/assets/11667761/06d86dfa-5a26-4ba5-b999-3c5e12cfb31e)
![Screenshot 2023-08-02 at 5 24 59 AM](https://github.com/ngageoint/mgrs-java/assets/11667761/acb79540-8a96-426e-bde1-df3592713632)

**Comment by tercjak** [comment]
I got those lines from this library and those  diagonal lines  are wrong so you have 3 bugs:
1) You generated wrong lines
2) You dont draw them to bitmap because of other bug .(so it posible that other lines also arent generated
3) You don't want to accept the error

when you are drawing overlay on screen your projection could be other than map projection.
Triangles are valid,
https://mappingsupport.com/p2/coordinates-mgrs-gissurfer-maps.html
also have those triangles so its ok

this is probably not the best resource if you are trying to learn how MGRS works
You are right but idk where is good resource ,According to https://github.com/patrykkalinowski/mgrs-wgs84/blob/master/mgrs.js each mgrs grid  Zone have bounds         var corners = [
            zone.name + column + row + coord_x + "0000" + coord_y + "0000",
            zone.name + column + row + coord_x + "0000" + coord_y + "9999",
            zone.name + column + row + coord_x + "9999" + coord_y + "9999",
            zone.name + column + row + coord_x + "9999" + coord_y + "0000",
            zone.name + column + row + coord_x + "0000" + coord_y + "0000",
          ] All libraries have this implemented diffriently with fixes for edge cases.

**Reply by bosborn** [reply]
Again, this library has only been used for tile generation.  The algorithm was ported from previous implementations.  It will probably take some effort on your part to adapt to using it for map lines.  If you feel there is a bug, you may
* Submit a test clearly demonstrating the issue both with expected and unexpected results
* Submit a pull request fixing the bug that does not break overlapping tile generations
* Submit a pull request that supports retrieving lines in a more appropriate for raw feature display

**Comment by tercjak** [comment]
I solved the problem by introducing value filtering on my side. If anyone in the future has a similar problem please contact me.

**Comment by spyhunter99** [comment]
@tercjak i have the same issue. what causes the extra diagonal line? did you find a way to filter it out?
